src/Craft/Systemd.hs

```diff
@@ -0,0 +1,384 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Craft.Systemd where
+
+import Control.Lens
+import Data.List.Utils (join)
+
+import Craft
+import Craft.File
+
+-- Helpful: https://www.digitalocean.com/community/tutorials/understanding-systemd-units-and-unit-files
+
+-- A service name is simply the name of that service
+type ServiceName = String
+
+-- All systemd Units can be transformed to their plaintext output
+class WritableUnit a where
+  transformUnit :: a -> String
+
+-- A generic implementation for the String type
+instance WritableUnit String where
+  transformUnit sn = show sn
+
+-- A generic implementation for the Bool type
+instance WritableUnit Bool where
+  transformUnit True = "true"
+  transformUnit False = "false"
+
+-- A generic implementation for the Int type
+instance WritableUnit Int where
+  transformUnit i = show i
+
+-- Composite units need to be able to furnish a filename (*without* a path)
+class CompositeUnit unit where
+  getFileName :: unit -> FilePath
+
+-- | Helpers |--
+-- =====================================================================
+-- Functions for extracting optional values to be transformed
+writeOptionalSection :: WritableUnit a => Maybe a -> String
+writeOptionalSection unit =
+  case unit of
+    {
+      Nothing -> "";
+      Just u -> transformUnit u ++ "\n"
+    }
+
+-- Optionally write a directive
+writeOptional :: WritableUnit a => String -> Maybe a -> String
+writeOptional directive optionalParam =
+  case optionalParam of
+    {
+      Nothing -> "";
+      Just param -> directive ++ "=" ++ transformUnit param ++ "\n"
+    }
+
+writeOptionalList :: WritableUnit a => String -> Maybe [a] -> String
+writeOptionalList directive optionalParams =
+  case optionalParams of
+    {
+      Nothing -> "";
+      Just params -> directive ++ "=" ++
+      (join " " (map transformUnit params)) ++ "\n"
+    }
+-- =====================================================================
+
+-- | [Unit] |--
+-- =====================================================================
+data UnitSection = UnitSection { _description   :: String
+                               , _documentation :: Maybe String
+                               , _requires      :: Maybe [ServiceName]
+                               , _wants         :: Maybe [ServiceName]
```

Comment:
> Would an empty `Requires=` be an error in systemd? Would Requires= mean something different than a n absent Requires field? If so, then its better to make it a list.

It wouldn't be an error, but this is just a subset of all the things systemd supports and I think it'd be much cleaner to not include those unless necessary. I will change it to be a list, and have an empty list not write something.

Any ideas how to express an optional string then? I'm not sure I want to use `""` to indicate the option is not set (but I guess I can if that's the best way to do it)

> It still means that it is optional, because `[]` is a valid value.

That's a good point, I guess the `Maybe` doesn't help too much then :-/